lib/Tools/circt-bmc/LowerToBMC.cpp

```diff
@@ -0,0 +1,191 @@
+  }
+
+  // Check that there's only one clock input to the module
+  // TODO: supporting multiple clocks isn't too hard, an interleaving of clock
+  // toggles just needs to be generated
+  bool hasClk = false;
+  for (auto input : hwModule.getInputTypes()) {
+    if (isa<seq::ClockType>(input)) {
+      if (hasClk) {
+        hwModule.emitError("designs with multiple clocks not yet supported");
+        return signalPassFailure();
+      }
+      hasClk = true;
+    }
+  }
+  {
+    OpBuilder::InsertionGuard guard(builder);
+    // Initialize clock to 0 if it exists, otherwise just yield nothing
+    auto *initBlock = builder.createBlock(&bmcOp.getInit());
+    builder.setInsertionPointToStart(initBlock);
+    if (hasClk) {
+      auto initVal =
+          builder.create<hw::ConstantOp>(loc, builder.getI1Type(), 0);
+      auto toClk = builder.create<seq::ToClockOp>(loc, initVal);
+      builder.create<verif::YieldOp>(loc, ValueRange{toClk});
+    } else {
+      builder.create<verif::YieldOp>(loc, ValueRange{});
+    }
+
+    // Toggle clock in loop region if it exists, otherwise just yield nothing
+    auto *loopBlock = builder.createBlock(&bmcOp.getLoop());
+    builder.setInsertionPointToStart(loopBlock);
+    if (hasClk) {
+      loopBlock->addArgument(seq::ClockType::get(ctx), loc);
+      auto fromClk =
+          builder.create<seq::FromClockOp>(loc, loopBlock->getArgument(0));
+      auto cNeg1 = builder.create<hw::ConstantOp>(loc, builder.getI1Type(), -1);
+      auto nClk = builder.create<comb::XorOp>(loc, fromClk, cNeg1);
+      auto toClk = builder.create<seq::ToClockOp>(loc, nClk);
+      // Only yield clock value
+      builder.create<verif::YieldOp>(loc, ValueRange{toClk});
+    } else {
+      builder.create<verif::YieldOp>(loc, ValueRange{});
+    }
+  }
+  bmcOp.getCircuit().takeBody(hwModule.getBody());
+  hwModule->erase();
+
+  // Define global string constants to print on success/failure
+  LLVM::GlobalOp successGlobal, failureGlobal;
+  {
+    OpBuilder::InsertionGuard guard(builder);
+    builder.setInsertionPointToEnd(getOperation().getBody());
+
+    StringRef successStr = "Bound reached with no violations!\n";
+    auto succArrayTy =
+        LLVM::LLVMArrayType::get(builder.getI8Type(), successStr.size() + 1);
+    successGlobal = builder.create<LLVM::GlobalOp>(
+        loc, succArrayTy, /*isConstant=*/true, LLVM::linkage::Linkage::Private,
+        successStr,
+        StringAttr::get(builder.getContext(),
+                        Twine(successStr).concat(Twine('\00'))));
+
+    StringRef failureStr = "Assertion can be violated!\n";
+    auto failArrayTy =
+        LLVM::LLVMArrayType::get(builder.getI8Type(), failureStr.size() + 1);
+    failureGlobal = builder.create<LLVM::GlobalOp>(
+        loc, failArrayTy, /*isConstant=*/true, LLVM::linkage::Linkage::Private,
+        failureStr,
+        StringAttr::get(builder.getContext(),
+                        Twine(failureStr).concat(Twine('\00'))));
```

Comment:
I didn't really mean to inline the whole function, sorry if that wasn't clear in my previous comment. Having the duplicate logic in a lambda would still be nice. I meant we could just generate a unique symbol and create new global ops each time instead of checking if a matching one already exists (i.e., not just use the content string as symbol).

Reply:
Ah my mistake! Yeah I misunderstood, I see your point now. I'll get that updated!

Reply:
@maerhart does the lambda version in the latest commit look good to you?

Reply:
Looks great! Thanks!